dali/python/nvidia/dali/pipeline.py

```diff
@@ -125,24 +140,38 @@ def __init__(self, batch_size = -1, num_threads = -1, device_id = -1, seed = -1,
         self._device_id = device_id
         self._seed = seed
         self._exec_pipelined = exec_pipelined
+        # When initializing DALI, we do the following in order:
+        # * Discover the ops specified in Python, group the ExternalSources
+        # * Start the Python workers pool
+        # * Construct the C++ Pipeline backend and pass the graph to it
```

Comment:
If this fails the pool is destroyed as well?

Reply:
You can list which method is responsible for each step.

Reply:
Pool is member of the pipeline and uses weakref.finalize to be cleaned up after the pipeline so we do not release memory to early (not sure how realistic is that pipeline will start being garbage collected and someone will still be using the no_copy memory to ExternalSource). It should be closed and released after pipeline is destroyed AFAIK. (look at _start_py_workers).